packages/registry/db/seeds/namespaceSeeder.ts

```diff
@@ -2,103 +2,292 @@ import { homedir } from 'os';
+  const namespaceRepo = queryRunner.manager.getRepository(Namespace);
+  await namespaceRepo.delete({ name: seedInput.name });
+
+  console.log(`deleted namespace ${seedInput.name}`);
 
-        map.set(service.name, { ...savedService, rotation: serviceRotation });
-        remaining--;
+  return { id: -1, name: seedInput.name, services: new Map() };
+};
 
-        console.log(`seeded ${service.name}`);
+const createNamespace = async (queryRunner: QueryRunner, seedInput: NamespaceSeederInput<ServiceCreateInput>): Promise<NamespaceSeederOutput> => {
+  const map: Map<string, Service & { rotation: number }> = new Map();
+  remaining = seedInput.services ? seedInput.services.length : 0;
+
+  // seed namespace
+  const namespaceRepo = queryRunner.manager.getRepository(Namespace);
+  const serviceRepo = queryRunner.manager.getRepository(Service);
+  const rotationRepo = queryRunner.manager.getRepository(Rotation);
+  const blockeRepo = queryRunner.manager.getRepository(Block);
+
+  const existingNamespace = await namespaceRepo.findOne({ where: { name: seedInput.name } });
+
+  if (existingNamespace !== null) {
+    console.warn(`namespace ${seedInput.name} already exists`);
+  }
+
+  const { namespaceId } = await namespaceRepo.save({ name: seedInput.name });
+
+  // seed services (the typeorm save method is used to seed the `service_closure` table) and their rotations
+  while (remaining > 0) {
+    for (const service of seedInput.services ?? []) {
+      if (map.has(service.name) || (service.parent !== undefined && !map.has(service.parent))) {
+        continue;
       }
+
+      const savedService = await serviceRepo.save({
+        namespaceId,
+        name: service.name,
+        parallelism: service.parallelism,
+        serviceType: service.serviceType,
+        parentServiceId: service.parent ?? null,
+        parent: service.parent !== undefined ? map.get(service.parent) : null,
+      });
+
+      const parent = seedInput.services?.find((s) => s.name === service.parent);
+
+      const serviceRotation = service.rotation ?? INITAL_ROTATION_VALUE;
+      await rotationRepo.save({
+        service: savedService,
+        serviceId: savedService.id,
+        serviceRotation,
+        parentRotation: parent !== undefined ? parent.rotation ?? INITAL_ROTATION_VALUE : null,
+      });
+
+      map.set(service.name, { ...savedService, rotation: serviceRotation });
+      remaining--;
+
+      console.log(`created service ${service.name}`);
     }
+  }
+
+  // seed blocks
+  for (const service of seedInput.services ?? []) {
+    const blockerService = map.get(service.name);
+    for (const blockee of service.blockees ?? []) {
+      const blockeeService = map.get(blockee);
+      await blockeRepo.save({
+        blockerId: (blockerService as Service).id,
+        blockerService: blockerService,
+        blockeeId: (blockeeService as Service).id,
```

Comment:
why do we use `as` here?>

Reply:
map.get() returns `map<T>` or undefined. at this point we know the return isn't undefined, I've refactored the return type to be `as Service` after calling map.get()